contrib/format-excel/src/main/java/org/apache/drill/exec/store/excel/ExcelBatchReader.java

```diff
@@ -0,0 +1,444 @@
+          case NUMERIC:
+            tempColumnName = String.valueOf(cell.getNumericCellValue());
+            makeColumn(builder, tempColumnName, TypeProtos.MinorType.FLOAT8);
+            excelFieldNames.add(colPosition, tempColumnName);
+            break;
+        }
+        colPosition++;
+      }
+    }
+    columnWriters = new ArrayList<ScalarWriter>(excelFieldNames.size());
+    return builder.buildSchema();
+  }
+
+  /**
+   * Helper function to get the selected sheet from the configuration
+   *
+   * @return XSSFSheet The selected sheet
+   */
+  private XSSFSheet getSheet() {
+    int sheetIndex = 0;
+    if (!readerConfig.sheetName.isEmpty()) {
+      sheetIndex = workbook.getSheetIndex(readerConfig.sheetName);
+    }
+
+    //If the sheet name is not valid, throw user exception
+    if (sheetIndex == -1) {
+      throw UserException
+        .validationError()
+        .message("Could not open sheet " + readerConfig.sheetName)
+        .build(logger);
+    } else {
+      return workbook.getSheetAt(sheetIndex);
+    }
+  }
+
+  @Override
+  public boolean next() {
+    recordCount = 0;
+    while (!rowWriter.isFull()) {
+      if (!nextLine(rowWriter)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  public boolean nextLine(RowSetLoader rowWriter) {
+    if (!rowIterator.hasNext()) {
+      return false;
+    } else if (recordCount >= readerConfig.lastRow) {
+      return false;
+    }
+
+    int lastRow = readerConfig.lastRow;
+    while (recordCount < lastRow && rowIterator.hasNext()) {
+
+      lineCount++;
+
+      Row row = rowIterator.next();
+      // If the user specified that there are no headers, get the column count
+      if (readerConfig.headerRow == -1 && recordCount == 0) {
+        this.totalColumnCount = row.getLastCellNum();
+      }
+
+      String fieldName;
+      if (row.getLastCellNum() < totalColumnCount) {
+        throw UserException.dataReadError().message("Wrong number of columns in row: %d", row.getLastCellNum()).build(logger);
+      }
+      int colPosition = 0;
+      if (readerConfig.firstColumn != 0) {
+        colPosition = readerConfig.firstColumn - 1;
+      }
+
+      int finalColumn = totalColumnCount;
+      if (readerConfig.lastColumn != 0) {
+        finalColumn = readerConfig.lastColumn - 1;
+      }
+      rowWriter.start();
+      for (int colWriterIndex = 0; colPosition < finalColumn; colPosition++) {
+        Cell cell = row.getCell(colPosition);
+
+        /*if (readerConfig.firstColumn != 0) {
+          colWriterIndex += readerConfig.firstColumn - 1;
+        }*/
+
+        CellValue cellValue = evaluator.evaluate(cell);
+        if (cellValue == null) {
+          String fieldValue = "";
+          if (firstLine) {
+            addColumnToArray(rowWriter, excelFieldNames.get(colPosition), MinorType.VARCHAR);
+          }
+          columnWriters.get(colWriterIndex).setString(fieldValue);
+
+        } else {
+          switch (cellValue.getCellTypeEnum()) {
```

Comment:
Wouldn't this still require some switch statement to load the data?  I see where you're going with this, but you still need to have a switch statement to even get the cell value as the ways of retrieving it are different.

Reply:
Hi @paul-rogers 
I see where you're going with this, however, there are a few issues and I'm not sure that it's going to be possible to remove the switch statement.  The main issue is that the column names are created by reading the header line.  This is done here:
https://github.com/apache/drill/blob/a23ca66bc5da2e89b3957ad0b3cabb8c4ebdfcc4/contrib/format-excel/src/main/java/org/apache/drill/exec/store/excel/ExcelBatchReader.java#L166-L237.

However, the actual data types aren't assigned until the reader actually reads the first line of data which occurs here:
https://github.com/apache/drill/blob/a23ca66bc5da2e89b3957ad0b3cabb8c4ebdfcc4/contrib/format-excel/src/main/java/org/apache/drill/exec/store/excel/ExcelBatchReader.java#L320-#L330

There has to be additional logic in the second part so that the reader can add the columns to the schema (see below) and that requires the `switch` statement because the ways of accessing the different columns from the Excel sheet are different.

https://github.com/apache/drill/blob/a23ca66bc5da2e89b3957ad0b3cabb8c4ebdfcc4/contrib/format-excel/src/main/java/org/apache/drill/exec/store/excel/ExcelBatchReader.java#L332-L356

If the columns were already defined, what you are describing makes sense, but I'm not seeing how to do it with getting rid of the `switch` statement.